libsql-sqlite3/src/vectorfloat32.c

```diff
@@ -0,0 +1,238 @@
+** copies or substantial portions of the Software.
+**
+** THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+** IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
+** FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
+** COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
+** IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+** CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+**
+******************************************************************************
+**
+** 32-bit floating point vector format utilities.
+*/
+#ifndef SQLITE_OMIT_VECTOR
+#include "sqliteInt.h"
+
+#include "vectorInt.h"
+
+#include <math.h>
+
+/**************************************************************************
+** Utility routines for debugging
+**************************************************************************/
+
+void vectorF32Dump(const Vector *pVec){
+  float *elems = pVec->data;
+  unsigned i;
+
+  assert( pVec->type == VECTOR_TYPE_FLOAT32 );
+
+  for(i = 0; i < pVec->dims; i++){
+    printf("%f ", elems[i]);
+  }
+  printf("\n");
+}
+
+/**************************************************************************
+** Utility routines for vector serialization and deserialization
+**************************************************************************/
+
+static inline unsigned formatF32(float value, char *pBuf, int nBufSize){
+  sqlite3_snprintf(nBufSize, pBuf, "%g", (double)value);
+  return strlen(pBuf);
+}
+
+static inline unsigned serializeF32(unsigned char *pBuf, float value){
+  u32 *p = (u32 *)&value;
+  pBuf[0] = *p & 0xFF;
+  pBuf[1] = (*p >> 8) & 0xFF;
+  pBuf[2] = (*p >> 16) & 0xFF;
+  pBuf[3] = (*p >> 24) & 0xFF;
+  return sizeof(float);
+}
+
+static inline float deserializeF32(const unsigned char *pBuf){
+  u32 value = 0;
+  value |= (u32)pBuf[0];
+  value |= (u32)pBuf[1] << 8;
+  value |= (u32)pBuf[2] << 16;
+  value |= (u32)pBuf[3] << 24;
+  return *(float *)&value;
+}
+
+size_t vectorF32SerializeToBlob(
+  const Vector *pVector,
+  unsigned char *pBlob,
+  size_t nBlobSize
+){
+  float *elems = pVector->data;
+  unsigned char *pPtr = pBlob;
+  size_t len = 0;
+  unsigned i;
+
+  assert( pVector->type == VECTOR_TYPE_FLOAT32 );
+  assert( pVector->dims <= MAX_VECTOR_SZ );
+  assert( nBlobSize >= pVector->dims * sizeof(float) );
+
+  for(i = 0; i < pVector->dims; i++){
+    pPtr += serializeF32(pPtr, elems[i]);
+    len += sizeof(float);
+  }
+  return len;
+}
+
+size_t vectorF32DeserializeFromBlob(
+  Vector *pVector,
+  const unsigned char *pBlob,
+  size_t nBlobSize
+){
+  float *elems = pVector->data;
+  unsigned i;
+  pVector->type = VECTOR_TYPE_FLOAT32;
+  pVector->dims = nBlobSize / sizeof(float);
+
+  assert( pVector->dims <= MAX_VECTOR_SZ );
+  assert( nBlobSize >= vectorDataSize(pVector->type, pVector->dims) );
```

Comment:
Isn't this redundant when 3 lines above we set pVector->dims to nBlobSize / sizeof(float)?